refactor(web): OSK spacebar-label updates now managed by layer object 🪠

jahorton opened the pull request:
To prepare to optimize layout-reflow for Web's OSK, it will help us to rework how spacebar-text and its styling are managed.  They'll be far easier to "batch" with other layout updates if managed by the layer containing the key, rather than set outside of the standard flow.

Also, we had a _lot_ of redundant calls to the `showLanguage()` function, which is what updated the spacebar.  We only need one.

This PR has been spun off of an older state of #11140.  Certain design decisions may make more sense with the whole picture in mind, though I have tried to recontextualize them to make better sense in isolation for this PR.

## User Testing

TEST_ANDROID_SPACEBAR:  Using Keyman for Android, verify that the spacebar is always appropriately captioned (based on user settings in the menu) for the active keyboard, with the text properly sized.

- If not already installed, download and test using `balochi_phonetic` - it has very high font scaling that will help stress-test part of spacebar-display management.
    - If the text overflows the spacebar at any point, overrunning its boundaries, FAIL this test.
- Test with a few other keyboards as well.
- Be sure to swap the device orientation back and forth a few times - from portrait to landscape and back again.
- Be sure to swap layers and verify that the spacebar for each is always labeled appropriately.

bharanidharanj commented:
# Test Results

* **TEST_ANDROID_SPACEBAR (FAILED):** Tested with the attached PR build (Keyman 17.0.0303-beta-test-11175 on different Android versions (Android 5.0, 7.1, 9.0 and 12.0, 13, 14) and here is my observation: 1. Installed balochi_phonetic keyboard. 2. Installed Khmer Angkor and SIL (IPA) keyboard. 3. Switched to  Balochi_Phonetic keyboard. Verified that the Spacebar captioned appropriately for the active keyboard. 4. Changed the orientation from portrait to landscape. 5. I observed that the label name is exceeding its boundaries in Android 5, 9, 7.1, and 9 OS versions. However, it displays correctly in Android 12, 13, and 14 OS versions.

..Android 5

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/d13c379a-d23a-4eae-84b4-62fa33a8ad99">

..Android 7.1

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/fa43be1b-9dc0-4d0c-a717-9c84325d72b2">

..Android 9.0

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/dd969808-6833-4609-b6d6-e70b98612ada">

..Android 12

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/7a8513fe-d0b0-45ce-bc4c-3e323c534a70">

jahorton commented:
> # Test Results
> 
> * **TEST_ANDROID_SPACEBAR (FAILED):** Tested with the attached PR build (Keyman 17.0.0303-beta-test-11175 on different Android versions (Android 5.0, 7.1, 9.0 and 12.0, 13, 14) and here is my observation: 1. Installed balochi_phonetic keyboard. 2. Installed Khmer Angkor and SIL (IPA) keyboard. 3. Switched to  Balochi_Phonetic keyboard. Verified that the Spacebar captioned appropriately for the active keyboard. 4. Changed the orientation from portrait to landscape. 5. I observed that the label name is exceeding its boundaries in Android 5, 9, 7.1, and 9 OS versions. However, it displays correctly in Android 12, 13, and 14 OS versions.
> 
> 
> ..Android 5
> <img alt="" width="300" src="https://private-user-images.githubusercontent.com/19683143/319971416-d13c379a-d23a-4eae-84b4-62fa33a8ad99.JPG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI1MzgwMzgsIm5iZiI6MTcxMjUzNzczOCwicGF0aCI6Ii8xOTY4MzE0My8zMTk5NzE0MTYtZDEzYzM3OWEtZDIzYS00ZWFlLTg0YjQtNjJmYTMzYThhZDk5LkpQRz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA4VDAwNTUzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxOTFjOGViZDQ0MzJmODEyZjYwZjcwMWRlMWYwYjdjYzQ4ZDBhNDdjYTVhNzQzYTZlMjA3ODdjODZlY2QyOGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WK-f5yblct_rycGo-eyUciR5Eo9dR4H2fFF8RBk4c7w">
> 
> [...]
> 
> ..Android 12
> <img alt="" width="300" src="https://private-user-images.githubusercontent.com/19683143/319971489-7a8513fe-d0b0-45ce-bc4c-3e323c534a70.JPG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI1MzgwMzgsIm5iZiI6MTcxMjUzNzczOCwicGF0aCI6Ii8xOTY4MzE0My8zMTk5NzE0ODktN2E4NTEzZmUtZDBiMC00NWNlLWJjNGMtM2UzMjNjNTM0YTcwLkpQRz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA4VDAwNTUzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4NWY3YzI4ZTNiNzdlZGUyZmMyN2I2NDQ4MGIxNjY0YWNmYmIwMjU5YTM1NGQ2YzAyNjk1MzIwNzljMzhlMDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gypsUa6eWtegJYwfrYF-iZTW8fMwmdV1aCWh9rRxpR8">

I'm going to call this a pass, but I _really_ can't blame you for reporting it this way.  TL,DR: technical limitations with older Chrome versions.

TEST_ANDROID_SPACEBAR:  PASS

----

The reason:  the technique we use to ensure that the key height stays within the spacebar's boundary isn't available to us in old versions of Chrome.  Per https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics#browser_compatibility, we need a minimum of Chrome 87 in order to scale based on key-height.  (See the [`fontBoundingBoxAscent`](https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics/fontBoundingBoxAscent) entry.)

Re: Android 7.1.1, I found a link suggesting that Android 7 shipped with Chrome 58.  For Android 9.0, APKMirror offers Chrome 73.  As Android Studio emulators typically start out with the minimum possible Chrome version, this explains why the height scaling wasn't working.

So, a couple of things of note:
1. No errors or crashes resulted from trying to use Web features not available given the version of Chrome in place.
2. The key _width_ scaling still worked perfectly fine.

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 17.0.305-beta](https://keyman.com/downloads/releases/beta/17.0.305)

mcdurdin commented on web/src/engine/osk/src/keyboard-layout/oskLayer.ts:
```diff
@@ -76,6 +81,21 @@ export default class OSKLayer {
+      const spacebarLabel = this.spaceBarKey.label;
+      let tButton = this.spaceBarKey.btn;
+
+      if (typeof (tButton.className) == 'undefined' || tButton.className == '') {
+        tButton.className = 'kmw-spacebar';
+      } else if (tButton.className.indexOf('kmw-spacebar') == -1) {
+        tButton.className += ' kmw-spacebar';
+      }
```
```suggestion
      tButton.classList.add('kmw-spacebar');
```

https://developer.mozilla.org/en-US/docs/Web/API/Element/classList

mcdurdin commented on web/src/engine/osk/src/keyboard-layout/oskLayer.ts:
```diff
@@ -76,6 +81,21 @@ export default class OSKLayer {
     this.scrollKey   = this.findKey('K_SCROLL');
+
+    if(this.spaceBarKey) {
+      const spacebarLabel = this.spaceBarKey.label;
+      let tButton = this.spaceBarKey.btn;
+
+      if (typeof (tButton.className) == 'undefined' || tButton.className == '') {
+        tButton.className = 'kmw-spacebar';
+      } else if (tButton.className.indexOf('kmw-spacebar') == -1) {
+        tButton.className += ' kmw-spacebar';
+      }
+
+      if (spacebarLabel.className != 'kmw-spacebar-caption') {
+        spacebarLabel.className = 'kmw-spacebar-caption';
+      }
```
```suggestion
      spacebarLabel.className = 'kmw-spacebar-caption';
```

No need to guard this ... I am sure it'll already be optimized.

mcdurdin replied:
Or how about this ... so much shorter:

```suggestion
      this.spaceBarKey.btn.classList.add('kmw-spacebar');
      this.spaceBarKey.label.className = 'kmw-spacebar-caption';
```

mcdurdin commented on web/src/engine/osk/src/keyboard-layout/oskLayer.ts:
```diff
@@ -96,7 +116,35 @@ export default class OSKLayer {
+   **/
+  showLanguage(displayName: string) {
+    if(!this.spaceBarKey) {
+      return () => {};
```
What's this about? Nothing else returns anything?

mcdurdin commented on web/src/engine/osk/src/keyboard-layout/oskLayer.ts:
```diff
@@ -96,7 +116,35 @@ export default class OSKLayer {
+        spacebarLabel.innerText = displayName;
+      }
+    }
+    catch (ex) { }
```
Why are we catching everything here?

mcdurdin commented on web/src/engine/osk/src/keyboard-layout/oskLayer.ts:
```diff
@@ -96,7 +116,35 @@ export default class OSKLayer {
+  }
+
+  public refreshLayout(vkbd: VisualKeyboard, layoutParams: LayerLayoutParams) {
+    const layerHeight = layoutParams.keyboardHeight;
```
Why is this here?

mcdurdin reviewed (commented):
This is pretty much fine, but there are some funny leftover bits and pieces and some code that really needs simplification.
